src/legacy/core_plugins/expressions/public/np_ready/public/expression_renderer.tsx

```diff
@@ -17,58 +17,117 @@
  * under the License.
  */
 
-import { useRef, useEffect } from 'react';
+import { useRef, useEffect, useState } from 'react';
 import React from 'react';
-import { ExpressionAST, IExpressionLoaderParams, IInterpreterResult } from './types';
-import { IExpressionLoader, ExpressionLoader } from './loader';
+import classNames from 'classnames';
+import { EuiLoadingChart, EuiProgress } from '@elastic/eui';
+import { ExpressionAST, IExpressionLoaderParams, IInterpreterErrorResult } from './types';
+import { ExpressionLoader } from './loader';
 
 // Accept all options of the runner as props except for the
 // dom element which is provided by the component itself
 export interface ExpressionRendererProps extends IExpressionLoaderParams {
   className: string;
   dataAttrs?: string[];
   expression: string | ExpressionAST;
-  /**
-   * If an element is specified, but the response of the expression run can't be rendered
-   * because it isn't a valid response or the specified renderer isn't available,
-   * this callback is called with the given result.
-   */
-  onRenderFailure?: (result: IInterpreterResult) => void;
+  renderError?: (error?: string | null) => React.ReactElement | React.ReactElement[];
+}
+
+interface State {
+  isEmpty: boolean;
+  isLoading: boolean;
+  error: null | Error;
 }
 
 export type ExpressionRenderer = React.FC<ExpressionRendererProps>;
 
-export const createRenderer = (loader: IExpressionLoader): ExpressionRenderer => ({
+const defaultState: State = {
+  isEmpty: true,
+  isLoading: false,
+  error: null,
+};
+
+export const ExpressionRendererImplementation = ({
   className,
   dataAttrs,
   expression,
-  onRenderFailure,
   ...options
 }: ExpressionRendererProps) => {
   const mountpoint: React.MutableRefObject<null | HTMLDivElement> = useRef(null);
   const handlerRef: React.MutableRefObject<null | ExpressionLoader> = useRef(null);
+  const [state, setState] = useState<State>({ ...defaultState });
 
+  // Re-fetch data automatically when the inputs change
   useEffect(() => {
-    if (mountpoint.current) {
-      if (!handlerRef.current) {
-        handlerRef.current = loader(mountpoint.current, expression, options);
-      } else {
-        handlerRef.current.update(expression, options);
-      }
-      handlerRef.current.data$.toPromise().catch(result => {
-        if (onRenderFailure) {
-          onRenderFailure(result);
-        }
-      });
+    if (handlerRef.current) {
+      handlerRef.current.update(expression, options);
```

Comment:
`options` seem to could contain `.renderError` key here, which `.update` function does not expect.

Reply:
There's a couple weird things about how we're handling params right now. I've fixed this as best as I could, and it allows us to use searchContext in dashboards now. There's still some behavioral questions I have here but we can solve this separately.